File: dns_lookup.py


### 1. `dns_lookup.py` (Python)

```python
# dns_lookup.py — Python версия

import dns.resolver
import dns.reversename
import json
import csv
import sys
import os
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime
import argparse
import ipaddress

class DNSLookupIPv6:
    def __init__(self, timeout=5, threads=20):
        self.timeout = timeout
        self.threads = threads
        self.resolver = dns.resolver.Resolver()
        self.resolver.timeout = timeout
        self.resolver.lifetime = timeout
        self.results = []

    def lookup_aaaa(self, domain):
        """Выполняет lookup AAAA записи для домена."""
        try:
            answers = self.resolver.resolve(domain, 'AAAA')
            return [str(r) for r in answers]
        except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer, dns.resolver.Timeout):
            return []
        except Exception:
            return []

    def lookup_ptr(self, ipv6_addr):
        """Выполняет reverse lookup для IPv6-адреса."""
        try:
            rev = dns.reversename.from_address(ipv6_addr)
            answers = self.resolver.resolve(rev, 'PTR')
            return [str(r) for r in answers]
        except:
            return []

    def is_ipv6(self, addr):
        """Проверяет, является ли строка IPv6-адресом."""
        try:
            ipaddress.IPv6Address(addr)
            return True
        except:
            return False

    def lookup_domain(self, domain):
        """Выполняет полный lookup для домена или IP."""
        domain = domain.strip()
        if not domain:
            return None

        # Проверяем, является ли домен IPv6-адресом
        if self.is_ipv6(domain):
            # Для IP-адреса делаем reverse lookup
            ptr = self.lookup_ptr(domain)
            result = {
                'domain': domain,
                'timestamp': datetime.now().isoformat(),
                'AAAA': [],
                'PTR': ptr if ptr else ['не найден']
            }
            return result

        # Для домена делаем AAAA lookup
        aaaa = self.lookup_aaaa(domain)
        result = {
            'domain': domain,
            'timestamp': datetime.now().isoformat(),
            'AAAA': aaaa,
            'PTR': []
        }

        # Reverse lookup для каждого AAAA адреса
        if aaaa:
            for ip in aaaa:
                ptr = self.lookup_ptr(ip)
                result['PTR'].extend(ptr if ptr else ['не найден'])
        else:
            result['PTR'] = ['не найден']

        return result

    def lookup_batch(self, domains):
        """Пакетная проверка с многопоточностью."""
        print(f"🔍 Выполняем lookup для {len(domains)} доменов...")
        start_time = time.time()

        with ThreadPoolExecutor(max_workers=self.threads) as executor:
            future_to_domain = {executor.submit(self.lookup_domain, d): d for d in domains if d.strip()}
            for future in as_completed(future_to_domain):
                domain = future_to_domain[future]
                try:
                    result = future.result()
                    if result:
                        self.results.append(result)
                        print(f"✅ {domain} — выполнено")
                    else:
                        print(f"⚠️ {domain} — ошибка")
                except Exception as e:
                    print(f"❌ {domain} — {e}")

        elapsed = time.time() - start_time
        print(f"⏱️ Время выполнения: {elapsed:.2f} сек.")
        return self.results

    def save_json(self, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(self.results, f, indent=2, ensure_ascii=False)
        print(f"💾 Сохранено JSON: {filename}")

    def save_csv(self, filename):
        if not self.results:
            print("⚠️ Нет данных для сохранения.")
            return
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["Domain", "Timestamp", "AAAA", "PTR"])
            for r in self.results:
                writer.writerow([
                    r['domain'],
                    r['timestamp'],
                    '; '.join(r['AAAA']) if r['AAAA'] else '—',
                    '; '.join(r['PTR']) if r['PTR'] else 'не найден'
                ])
        print(f"💾 Сохранено CSV: {filename}")

    def print_table(self):
        if not self.results:
            print("Нет результатов.")
            return

        print("\n" + "=" * 90)
        print(f"{'Домен/IP':<25} {'AAAA':<35} {'PTR':<25}")
        print("-" * 90)
        for r in self.results:
            aaaa = '; '.join(r['AAAA']) if r['AAAA'] else '—'
            ptr = '; '.join(r['PTR']) if r['PTR'] else 'не найден'
            # Обрезаем длинные строки
            if len(aaaa) > 35:
                aaaa = aaaa[:35] + '...'
            if len(ptr) > 25:
                ptr = ptr[:25] + '...'
            print(f"{r['domain']:<25} {aaaa:<35} {ptr:<25}")
        print("=" * 90)

def main():
    parser = argparse.ArgumentParser(description='DNS Lookup IPv6 Pro')
    parser.add_argument('input', help='Файл с доменами/IP (по одному на строку)')
    parser.add_argument('-o', '--output', default='dns_results', help='Базовое имя файлов для сохранения')
    parser.add_argument('-t', '--threads', type=int, default=20, help='Количество потоков')
    parser.add_argument('--timeout', type=int, default=5, help='Таймаут запроса (сек)')
    parser.add_argument('--no-save', action='store_true', help='Не сохранять результаты')
    args = parser.parse_args()

    if not os.path.exists(args.input):
        print(f"❌ Файл {args.input} не найден.")
        sys.exit(1)

    print("🌐 DNS Lookup IPv6 Pro (Python)")
    with open(args.input, 'r', encoding='utf-8') as f:
        domains = [line.strip() for line in f if line.strip()]

    if not domains:
        print("❌ Нет доменов для проверки.")
        sys.exit(1)

    print(f"📂 Загружено {len(domains)} доменов/IP.")

    lookup = DNSLookupIPv6(timeout=args.timeout, threads=args.threads)
    results = lookup.lookup_batch(domains)

    if results:
        lookup.print_table()
        if not args.no_save:
            lookup.save_json(args.output + '.json')
            lookup.save_csv(args.output + '.csv')
    else:
        print("❌ Не удалось выполнить lookup ни для одного домена.")

if __name__ == "__main__":
    main()

File: dns_lookup.rb
# dns_lookup.rb — Ruby версия

require 'resolv'
require 'json'
require 'csv'
require 'time'
require 'ipaddr'
require 'concurrent'

class DNSLookupIPv6
  attr_reader :results

  def initialize(timeout: 5, threads: 20)
    @timeout = timeout
    @threads = threads
    @results = []
    @resolver = Resolv::DNS.new
    @resolver.timeouts = timeout
  end

  def is_ipv6?(addr)
    begin
      IPAddr.new(addr).ipv6?
    rescue
      false
    end
  end

  def lookup_aaaa(domain)
    begin
      @resolver.getresources(domain, Resolv::DNS::Resource::IN::AAAA)
               .map { |r| r.address.to_s }
    rescue Resolv::ResolvError, Resolv::ResolvTimeout
      []
    end
  end

  def lookup_ptr(ip)
    begin
      @resolver.getnames(ip).map(&:to_s)
    rescue
      []
    end
  end

  def lookup_domain(domain)
    domain = domain.strip
    return nil if domain.empty?

    # Проверяем, является ли строка IPv6-адресом
    if is_ipv6?(domain)
      ptr = lookup_ptr(domain)
      return {
        domain: domain,
        timestamp: Time.now.iso8601,
        AAAA: [],
        PTR: ptr.empty? ? ['не найден'] : ptr
      }
    end

    aaaa = lookup_aaaa(domain)
    ptr = []
    if aaaa.any?
      aaaa.each do |ip|
        ptr.concat(lookup_ptr(ip))
      end
    end
    ptr = ['не найден'] if ptr.empty?

    {
      domain: domain,
      timestamp: Time.now.iso8601,
      AAAA: aaaa,
      PTR: ptr
    }
  end

  def lookup_batch(domains)
    puts "🔍 Выполняем lookup для #{domains.size} доменов..."
    start = Time.now

    pool = Concurrent::FixedThreadPool.new(@threads)
    futures = domains.map do |domain|
      Concurrent::Future.execute(executor: pool) do
        result = lookup_domain(domain)
        if result
          @results << result
          puts "✅ #{domain} — выполнено"
        end
      end
    end
    futures.each(&:wait)

    elapsed = Time.now - start
    puts "⏱️ Время выполнения: #{elapsed.round(2)} сек."
    @results
  end

  def save_json(filename)
    File.write(filename, JSON.pretty_generate(@results))
    puts "💾 Сохранено JSON: #{filename}"
  end

  def save_csv(filename)
    return if @results.empty?
    CSV.open(filename, 'w') do |csv|
      csv << ['Domain', 'Timestamp', 'AAAA', 'PTR']
      @results.each do |r|
        csv << [
          r[:domain],
          r[:timestamp],
          r[:AAAA].join('; '),
          r[:PTR].join('; ')
        ]
      end
    end
    puts "💾 Сохранено CSV: #{filename}"
  end

  def print_table
    return if @results.empty?

    puts "\n" + "=" * 90
    printf "%-25s %-35s %-25s\n", "Домен/IP", "AAAA", "PTR"
    puts "-" * 90
    @results.each do |r|
      aaaa = r[:AAAA].join('; ')
      aaaa = '—' if aaaa.empty?
      aaaa = aaaa[0...35] + '...' if aaaa.length > 35
      ptr = r[:PTR].join('; ')
      ptr = ptr[0...25] + '...' if ptr.length > 25
      printf "%-25s %-35s %-25s\n", r[:domain], aaaa, ptr
    end
    puts "=" * 90
  end
end

def main
  if ARGV.empty?
    puts "Usage: ruby dns_lookup.rb <domains.txt>"
    puts "   или: ruby dns_lookup.rb domain1.com domain2.com"
    exit 1
  end

  domains = []

  if ARGV.size == 1 && File.exist?(ARGV[0])
    domains = File.readlines(ARGV[0]).map(&:strip).reject(&:empty?)
  else
    domains = ARGV
  end

  if domains.empty?
    puts "❌ Нет доменов для проверки."
    exit 1
  end

  puts "🌐 DNS Lookup IPv6 Pro (Ruby)"
  puts "📂 Загружено #{domains.size} доменов."

  lookup = DNSLookupIPv6.new(timeout: 5, threads: 20)
  lookup.lookup_batch(domains)

  unless lookup.results.empty?
    lookup.print_table
    lookup.save_json('dns_results.json')
    lookup.save_csv('dns_results.csv')
  else
    puts "❌ Не удалось выполнить lookup."
  end
end

main if __FILE__ == $0
